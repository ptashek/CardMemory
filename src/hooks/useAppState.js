// @flow
import React from 'react';
import colors from '@workday/canvas-colors-web';

type ClickActionPayload = {
  id: number,
  value: $Keys<typeof colors>,
};

type Action = {
  type: string,
  payload?: ClickActionPayload,
};

export type State = {
  selected: Map<number, string>,
  solved: Set<string>,
  moves: number,
  timer: number,
};

export type Dispatch = (action: Action) => void;

const initialState = (): State => ({
  selected: new Map(),
  solved: new Set(),
  moves: 0,
  timer: 0,
});

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    default:
      return state;

    case 'timer':
      return { ...state, timer: action.timer };

    case 'restart':
      return initialState();

    case 'clearSelected': {
      const nextState = { ...state };
      nextState.selected.clear();
      return nextState;
    }

    case 'click': {
      if (!action?.payload || state.selected.size === 2) {
        return state;
      }

      const { id, value } = action.payload;
      const nextState: State = { ...state };

      if (!nextState.selected.has(id)) {
        nextState.selected.set(id, value);
        const entries = nextState.selected.entries();
        const matchedSelectedEntries = Array.from(entries).filter(([_, v]) => v === value);

        if (matchedSelectedEntries.length === 2) {
          matchedSelectedEntries.forEach(([key]) => {
            nextState.selected.delete(key);
          });
          nextState.solved.add(value);
        }
      } else {
        nextState.selected.delete(id);
      }

      // each click represents half a move
      nextState.moves += 0.5;
      return nextState;
    }
  }
};

export default (): [State, Dispatch] => React.useReducer(reducer, initialState());
