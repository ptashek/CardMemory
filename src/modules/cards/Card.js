// @flow
import type { ComponentType } from 'react';
import React from 'react';
import classnames from 'classnames';
import colors from '@workday/canvas-colors-web';
import { makeStyles } from '@material-ui/core/styles';
import Avatar from '@material-ui/core/Avatar';
import Box from '@material-ui/core/Box';
import HelpOutlineIcon from '@material-ui/icons/HelpOutline';
import styles from './styles/Card';

const useStyles = makeStyles(styles);

type CardProps = {
  faceColor: $Keys<typeof colors>,
  faceText: string,
  faceUp: boolean,
  onClick: () => void,
};

type CardFaceProps = {
  color: $Keys<typeof colors>,
  text: string,
};

type CardElement = ComponentType<CardProps>;

const CardBack: ComponentType<{}> = () => {
  const classes = useStyles();

  return (
    <HelpOutlineIcon
      className={classnames(classes.cardContent, classes.faceCommon, classes.icon)}
    />
  );
};

export const CardFace: ComponentType<CardFaceProps> = (props: CardFaceProps) => {
  const classes = useStyles();

  return (
    <Box
      borderRadius={16}
      bgcolor={`colors.${props.color}`}
      width="100%"
      height="100%"
      className={classnames(classes.cardContent, classes.faceUp)}
    >
      <Avatar component="span" className={classnames(classes.faceCommon, classes.avatar)}>
        {props.text}
      </Avatar>
    </Box>
  );
};

const Card: CardElement = (props: CardProps) => {
  const classes = useStyles();
  let content;

  if (props.faceUp) {
    content = <CardFace text={props.faceText} color={props.faceColor} />;
  } else {
    content = <CardBack />;
  }

  return (
    <Box
      borderRadius={16}
      boxShadow={3}
      width={80}
      height={80}
      m={2}
      onClick={props.onClick}
      className={classnames(classes.card, { [classes.faceUp]: props.faceUp })}
    >
      {content}
    </Box>
  );
};

export default React.memo<CardProps>(Card);
