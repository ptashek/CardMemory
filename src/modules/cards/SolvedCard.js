// @flow
import type { ComponentType } from 'react';
import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Box from '@material-ui/core/Box';
import Avatar from '@material-ui/core/Avatar';
import colors from '@workday/canvas-colors-web';
import styles from './styles/SolvedCard';

const useStyles = makeStyles(styles);

type SolvedCardProps = {
  faceColor: $Keys<typeof colors>,
  faceText: string,
};

const SolvedCard: ComponentType<SolvedCardProps> = (props: SolvedCardProps) => {
  const classes = useStyles();

  return (
    <Box
      data-testid="solved-card"
      borderRadius={16}
      boxShadow={3}
      width={80}
      height={80}
      m={2}
      bgcolor={`colors.${props.faceColor}`}
    >
      <Avatar className={classes.avatar}>{props.faceText}</Avatar>
    </Box>
  );
};

export default React.memo<SolvedCardProps>(SolvedCard);
