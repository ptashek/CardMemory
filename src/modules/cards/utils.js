// @flow
import colors from '@workday/canvas-colors-web';

export type CardPair = [string, Number];

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Math/random
const getRandomInt = (rangeStart: number, rangeEnd: Number): Number => {
  const min = Math.ceil(rangeStart);
  const max = Math.floor(rangeEnd);
  return Math.floor(Math.random() * (max - min)) + min;
};

// https://en.wikipedia.org/wiki/Fisher-Yates_shuffle
const shuffle = <T>(input: Array<T>): Array<T> => {
  const output = [...input];
  const lastIndex = output.length - 1;
  let tmp, targetIndex;

  for (let sourceIndex = lastIndex; sourceIndex > 0; sourceIndex--) {
    targetIndex = getRandomInt(0, sourceIndex);
    tmp = output[sourceIndex];
    output[sourceIndex] = output[targetIndex];
    output[targetIndex] = tmp;
  }

  return output;
};

/* 
  only include the mid-range hues from the canvas palette
  https://design.workday.com/tokens/basic/colors/canvas-colors
*/
const randomColors: string[] = shuffle<string>(
  Object.keys(colors).filter((name: string) => ['200', '300', '400'].includes(name.substr(-3))),
);

export const generateCardPairs = (pairCount: Number): CardPair[] => {
  const cards: CardPair[] = [];
  const cardCount: Number = pairCount * 2;

  for (let i: Number = 0; i < cardCount; i++) {
    const pairIndex = Math.floor(i / 2);
    const colorName: string = randomColors[pairIndex];
    cards.push([colorName, pairIndex]);
  }

  return shuffle<CardPair>(cards);
};
