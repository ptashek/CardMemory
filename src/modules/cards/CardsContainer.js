// @flow
import type { ComponentType } from 'react';
import type { State, Dispatch } from '../../hooks/useAppState';
import type { CardPair } from '../../utils';
import React from 'react';
import Box from '@material-ui/core/Box';
import colors from '@workday/canvas-colors-web';
import AppStateContext from '../app/AppStateContext';
import { generateCardPairs } from '../../utils';
import SolvedCard from './SolvedCard';
import Card from './Card';

type CardElement = React$Element<typeof Card | typeof SolvedCard>;

type VoidCallback = () => void;
type ClickHandler = (id: number, value: $Keys<typeof colors>) => VoidCallback;

const CardsContainer: ComponentType<{}> = () => {
  const [state, dispatch, pairCount]: [State, Dispatch, number] = React.useContext(AppStateContext);

  React.useEffect(() => {
    if (state.selected.size === 2) {
      setTimeout(() => {
        dispatch({ type: 'clearSelected' });
      }, 1000);
    }
  }, [state, dispatch]);

  const clickHandlerFor: ClickHandler = React.useCallback(
    (id: number, value: $Keys<typeof colors>): VoidCallback => () => {
      dispatch({
        type: 'click',
        payload: {
          id,
          value,
        },
      });
    },
    [dispatch],
  );

  const cardPairs = React.useMemo<CardPair[]>(() => generateCardPairs(pairCount), [pairCount]);

  const cardElements = React.useMemo<CardElement[]>(() => {
    return cardPairs.map(([color, pairIndex], id) => {
      const cardKey = `card-${id}`;

      if (state.solved.has(color)) {
        return <SolvedCard faceColor={color} faceText={String(pairIndex + 1)} key={cardKey} />;
      }

      return (
        <Card
          faceUp={state.selected.has(id)}
          faceColor={color}
          faceText={String(pairIndex + 1)}
          onClick={clickHandlerFor(id, color)}
          key={cardKey}
        />
      );
    });
  }, [cardPairs, state, clickHandlerFor]);

  return (
    <Box display="flex" flexWrap="wrap" justifyContent="flex-start" p={2} m={2}>
      {cardElements}
    </Box>
  );
};

export default React.memo<{}>(CardsContainer);
