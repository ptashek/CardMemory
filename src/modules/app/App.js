// @flow
import type { ComponentType } from 'react';
import type { State, Dispatch } from '../../hooks/useAppState';
import React from 'react';
import { ThemeProvider } from '@material-ui/core/styles';
import CssBaseline from '@material-ui/core/CssBaseline';
import useAppState from '../../hooks/useAppState';
import CardsContainer from '../cards/CardsContainer';
import AppStateContext from './AppStateContext';
import AppBar from './AppBar';
import Stats from './Stats';
import GameOverModal from './GameOverModal';
import theme from './styles/theme';
import { generateCardPairs } from '../cards/utils';

// exported for testing purposes only
export const PAIR_COUNT: number = 10;

const App: ComponentType<{}> = () => {
  const [state, dispatch]: [State, Dispatch] = useAppState();
  const [cardPairs, setCardPairs] = React.useState(generateCardPairs(PAIR_COUNT));

  // dispatch is stable
  /* eslint-disable react-hooks/exhaustive-deps */
  const restartGame = React.useCallback(() => {
    dispatch({ type: 'restart' });
    setCardPairs(generateCardPairs(PAIR_COUNT));
  }, []);
  /* eslint-enable react-hooks/exhaustive-deps */

  return (
    <>
      <CssBaseline />
      <ThemeProvider theme={theme}>
        <AppStateContext.Provider value={[state, dispatch, PAIR_COUNT]}>
          <AppBar onRestartClick={restartGame} />
          <GameOverModal onPlayAgainClick={restartGame} />
          <Stats />
          <CardsContainer cardPairs={cardPairs} />
        </AppStateContext.Provider>
      </ThemeProvider>
    </>
  );
};

export default App;
