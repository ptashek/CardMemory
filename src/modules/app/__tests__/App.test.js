import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react';
import * as utils from '../../cards/utils';
import useAppState from '../../../hooks/useAppState';
import App, { PAIR_COUNT as mockPairCount } from '../App';

const mockDispatch = jest.fn();
const mockAppState = {
  selected: new Map([[1, '1']]),
  solved: new Set(),
  moves: 1,
  timer: 1,
};

jest.mock('../../../hooks/useAppState', () => {
  return jest.fn(() => [mockAppState, mockDispatch]);
});

describe('<App />', () => {
  beforeAll(() => {
    const cardCount = mockPairCount * 2;
    const cards = [];
    for (let i = 0; i < cardCount; i++) {
      const pairIndex = Math.floor(i / 2);
      const colorName = `color-${pairIndex}`;
      cards.push([colorName, pairIndex]);
    }
    utils.generateCardPairs = jest.fn().mockReturnValue(cards);
  });

  afterEach(() => {
    utils.generateCardPairs.mockClear();
  });

  afterAll(() => {
    utils.generateCardPairs.mockRestore();
  });

  test('matches snapshot', () => {
    const { container } = renderWithTheme(<App />);
    expect(container).toMatchSnapshot();
  });

  test('renders the app log', () => {
    const { getByTestId } = renderWithTheme(<App />);
    expect(getByTestId('app-logo')).toBeInTheDocument();
  });

  test('renders the app title', () => {
    const { getByText } = renderWithTheme(<App />);
    const titleElement = getByText('Card Memory');
    expect(titleElement).toBeInTheDocument();
  });

  describe('the restart button', () => {
    test('renders', () => {
      const { getByRole } = renderWithTheme(<App />);
      const buttonElement = getByRole('button');
      expect(buttonElement.textContent).toEqual('Restart');
    });

    test('when enabled and clicked regenerates the cards', () => {
      const { getByRole } = renderWithTheme(<App />);
      fireEvent.click(getByRole('button'));
      expect(mockDispatch).toHaveBeenCalledWith({ type: 'restart' });
      expect(utils.generateCardPairs).toHaveBeenNthCalledWith(2, mockPairCount);
    });
  });

  test('renders the intro text', () => {
    const { getByText } = renderWithTheme(<App />);
    let introElement = getByText(
      'Click each card to reveal it. When a pair is matched, it will remain visible until the game finishes. Only two unmatched cards can be revealed at any given time.',
    );

    expect(introElement).toBeInTheDocument();
  });

  test('renders the correct number of cards', () => {
    const { getAllByTestId } = renderWithTheme(<App />);
    const cardElements = getAllByTestId('card');
    expect(cardElements.length).toEqual(mockPairCount * 2);
  });
});
