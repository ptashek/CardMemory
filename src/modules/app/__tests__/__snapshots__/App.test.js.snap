// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> matches snapshot 1`] = `
<div>
  <header
    class="MuiPaper-root-15 MuiAppBar-root-4 MuiAppBar-positionSticky-7 MuiAppBar-colorPrimary-11 makeStyles-root-1 MuiPaper-elevation4-22"
  >
    <div
      class="MuiToolbar-root-43 MuiToolbar-dense-46 MuiToolbar-gutters-44"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-icon-3"
        data-testid="app-logo"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15 7.5V2H9v5.5l3 3 3-3zM7.5 9H2v6h5.5l3-3-3-3zM9 16.5V22h6v-5.5l-3-3-3 3zM16.5 9l-3 3 3 3H22V9h-5.5z"
        />
      </svg>
      <h6
        class="MuiTypography-root-56 makeStyles-title-2 MuiTypography-h6-66 MuiTypography-noWrap-75"
      >
        Card Memory
      </h6>
      <button
        class="MuiButtonBase-root-113 MuiButton-root-86 MuiButton-contained-94 MuiButton-containedSecondary-96"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label-87"
        >
          <span
            class="MuiButton-startIcon-108 MuiButton-iconSizeMedium-111"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root-47"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7.11 8.53L5.7 7.11C4.8 8.27 4.24 9.61 4.07 11h2.02c.14-.87.49-1.72 1.02-2.47zM6.09 13H4.07c.17 1.39.72 2.73 1.62 3.89l1.41-1.42c-.52-.75-.87-1.59-1.01-2.47zm1.01 5.32c1.16.9 2.51 1.44 3.9 1.61V17.9c-.87-.15-1.71-.49-2.46-1.03L7.1 18.32zM13 4.07V1L8.45 5.55 13 10V6.09c2.84.48 5 2.94 5 5.91s-2.16 5.43-5 5.91v2.02c3.95-.49 7-3.85 7-7.93s-3.05-7.44-7-7.93z"
              />
            </svg>
          </span>
          Restart
        </span>
        <span
          class="MuiTouchRipple-root-153"
        />
      </button>
    </div>
  </header>
  <h6
    class="MuiTypography-root-56 makeStyles-stats-116 MuiTypography-h6-66"
  >
    Click each card to reveal it. When a pair is matched, it will remain visible until the game finishes.
    <br />
    Unmatched pairs will be hidden after a short delay.
    <p />
    You have solved 
    0
     out of 
    10
     pairs in 
    1
     moves
  </h6>
  <div
    class="MuiBox-root-117 MuiBox-root-118"
  >
    <div
      class="MuiBox-root-117 MuiBox-root-125 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-126 makeStyles-card-119 makeStyles-faceUp-124"
      data-testid="card"
    >
      <div
        class="MuiBox-root-117 MuiBox-root-127 makeStyles-cardContent-123 makeStyles-faceUp-124"
      >
        <span
          class="MuiAvatar-root-128 MuiAvatar-circle-130 makeStyles-faceCommon-120 makeStyles-avatar-122 MuiAvatar-colorDefault-129"
        >
          1
        </span>
      </div>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-135 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-136 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-137 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-138 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-139 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-140 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-141 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-142 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-143 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-144 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-145 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-146 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-147 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-148 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-149 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-150 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-151 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
    <div
      class="MuiBox-root-117 MuiBox-root-152 makeStyles-card-119"
      data-testid="card"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root-47 makeStyles-cardContent-123 makeStyles-faceCommon-120 makeStyles-icon-121"
        data-testid="card-back-icon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11 18h2v-2h-2v2zm1-16C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8zm0-14c-2.21 0-4 1.79-4 4h2c0-1.1.9-2 2-2s2 .9 2 2c0 2-3 1.75-3 5h2c0-2.25 3-2.5 3-5 0-2.21-1.79-4-4-4z"
        />
      </svg>
    </div>
  </div>
</div>
`;
